docker/cli
# Fix hang on `docker stack deploy` with Kubernetes 1.14

There were 2 issues in this logic:
- we did watch for stack with non-matching label selector: we used the
one for getting stacks child resources, instead of a field selector on
the stack name
- something in kube 1.14 has changed the behavior of watches against
aggregated API that makes context cancellation effective only when the
next event is received. I don't know exactly where the problem is, but
it is certainly on the central API Server or in Kube-Proxy, as the watch
hangs with any version of go client vendoring with a server running 1.14

This should be back-ported to older version of the CLI

**- What I did**

I fixed the way we watch for stack events on stack deploy.
- our watch query was not matching the stack we just created
- with a server running Kubernetes 1.14, the watch hanged for 7 minutes (default watch timeout)

**- How I did it**

- For the watch query, I replaced the wrong label selector with a name field selector
- For the hang, the main issue seem to be a bug in kubernetes 1.14 were watches on aggregated resources are effective only when the server sends the next event. Combined with the other bug (where we did not match the actual stack we were supposed to watch), we had a systematic hang. With the previous fix, depending on timing, the watcher could hang as well, but not all time. I fixed this by intercepting the watcher interface to bypass graceful cancellation of the server-side watcher.

**- How to verify it**

- Deploy compose-on-kubernetes on a Kubernetes v1.14 cluster
- Deploy a stack on it

Note @thaJeztah: this should be backported to older CLI versions as well
@zappy-shu https://github.com/docker/cli/pull/1774 should be rebased on this one for full 1.14 compatibility.
